File: solutions/R/building_prediction_models_ex2.R
# Exercise 2: Missing data simulation
# Compare complete case analysis, single mean imputation, and multiple
# imputation on the simulated Framingham cohort from the chapter.

library(mice)

# --- The cohort, exactly as in the chapter ---------------------------------
set.seed(2024)
n <- 2000

framingham <- data.frame(
  age = round(runif(n, 30, 74)),
  male = rbinom(n, 1, 0.48),
  sbp = round(rnorm(n, 130, 18)),
  total_chol = round(rnorm(n, 210, 38)),
  hdl_chol = round(rnorm(n, 52, 15)),
  smoking = rbinom(n, 1, 0.22),
  diabetes = rbinom(n, 1, 0.08),
  bp_treatment = rbinom(n, 1, 0.15)
)

# The true coefficients. Because we simulated the data we know them, which is
# what makes this comparison possible at all.
truth <- c(
  "(Intercept)" = -7.5, age = 0.06, male = 0.4, sbp = 0.012,
  total_chol = 0.005, hdl_chol = -0.02, smoking = 0.5,
  diabetes = 0.7, bp_treatment = 0.3
)

lp <- with(
  framingham,
  -7.5 + 0.06 * age + 0.4 * male + 0.012 * sbp + 0.005 * total_chol -
    0.02 * hdl_chol + 0.5 * smoking + 0.7 * diabetes + 0.3 * bp_treatment
)
framingham$cvd_10yr <- rbinom(n, 1, plogis(lp))
cat("Cohort:", n, "patients,", sum(framingham$cvd_10yr), "events\n")

model_formula <- cvd_10yr ~ age + male + sbp + total_chol + hdl_chol +
  smoking + diabetes + bp_treatment

# --- Make two predictors missing, under MAR -------------------------------
# The mechanism matters more than the amount. Missingness in total_chol is made
# to depend on the OUTCOME and on age -- both recorded, so this is MAR, not
# MNAR. That choice is deliberate: if missingness depended only on the
# predictors already in the model, complete case analysis would still be
# unbiased for the coefficients, and there would be nothing to see.
punch_holes <- function(df, intercept) {
  set.seed(7)
  p_chol <- plogis(intercept + 1.0 * df$cvd_10yr + 0.03 * (df$age - 52))
  p_hdl <- plogis(intercept + 0.2 + 0.9 * df$smoking + 0.02 * (df$sbp - 130))
  df$total_chol[runif(nrow(df)) < p_chol] <- NA
  df$hdl_chol[runif(nrow(df)) < p_hdl] <- NA
  df
}

incomplete <- punch_holes(framingham, intercept = -2.2)

cat("Missing total_chol:", sum(is.na(incomplete$total_chol)), "\n")
cat("Missing hdl_chol: ", sum(is.na(incomplete$hdl_chol)), "\n")
cat(
  "Complete rows:", sum(complete.cases(incomplete)),
  sprintf("(%.0f%% of the cohort discarded by a complete case analysis)\n",
          100 * mean(!complete.cases(incomplete)))
)

# --- The three approaches, plus the full data as a benchmark --------------
fit_full <- glm(model_formula, data = framingham, family = binomial)
fit_cca <- glm(model_formula, data = incomplete, family = binomial)

mean_imputed <- incomplete
for (v in c("total_chol", "hdl_chol")) {
  mean_imputed[[v]][is.na(mean_imputed[[v]])] <- mean(mean_imputed[[v]], na.rm = TRUE)
}
fit_mean <- glm(model_formula, data = mean_imputed, family = binomial)

# 20 imputations, predictive mean matching, pooled with Rubin's rules
imp <- mice(incomplete, m = 20, method = "pmm", seed = 42, printFlag = FALSE)
pool_obj <- pool(with(
  imp,
  glm(cvd_10yr ~ age + male + sbp + total_chol + hdl_chol + smoking +
        diabetes + bp_treatment, family = binomial)
))
pooled <- summary(pool_obj)
mi_est <- setNames(pooled$estimate, pooled$term)
mi_se <- setNames(pooled$std.error, pooled$term)
# lambda is the share of the pooled variance that comes from the missing data
# (the between-imputation part). It is precisely what single imputation drops.
mi_lambda <- setNames(pool_obj$pooled$lambda, pool_obj$pooled$term)

# --- (a) Which approach lands closest to the truth? ----------------------
comparison <- data.frame(
  truth = truth,
  full_data = coef(fit_full),
  complete_case = coef(fit_cca),
  mean_imputation = coef(fit_mean),
  multiple_imputation = mi_est[names(truth)]
)
cat("\n--- Coefficient estimates ---\n")
print(round(comparison, 4))

# Two error measures, and the difference between them is the point. Distance
# from the truth mixes up two things: damage done by the missing data, and the
# sampling noise that was already in this cohort of 2000. Distance from the
# full-data estimates isolates the first.
slopes <- rownames(comparison) != "(Intercept)"
mae <- function(x) mean(abs(x - comparison$truth[slopes]))
mae_full <- function(x) mean(abs(x - comparison$full_data[slopes]))

cat("\n--- Mean absolute error across the 8 slopes ---\n")
err <- data.frame(
  vs_truth = sapply(comparison[slopes, -1], mae),
  vs_full_data = sapply(comparison[slopes, -1], mae_full)
)
print(round(err, 5))

cat("\nRanked by distance from the full-data estimates:\n")
print(round(sort(err$vs_full_data[-1] |> setNames(rownames(err)[-1])), 5))

# --- (b) What happens to the standard errors? ---------------------------
se_table <- data.frame(
  complete_case = summary(fit_cca)$coefficients[, "Std. Error"],
  mean_imputation = summary(fit_mean)$coefficients[, "Std. Error"],
  multiple_imputation = mi_se[names(truth)]
)
cat("\n--- Standard errors ---\n")
print(round(se_table, 5))

cat("\nMean imputation SE as a percentage of the multiple-imputation SE,\n")
cat("for the two variables that were actually imputed:\n")
for (v in c("total_chol", "hdl_chol")) {
  cat(sprintf(
    "  %-11s %.1f%%\n", v,
    100 * se_table[v, "mean_imputation"] / se_table[v, "multiple_imputation"]
  ))
}

# What single imputation actually discards, as a number. Rubin's rules split the
# pooled variance into a within-imputation part (ordinary sampling uncertainty)
# and a between-imputation part (uncertainty about the guesses themselves).
# Single imputation sets the second to zero by construction.
cat("\nShare of the pooled MI variance that comes from the imputation itself:\n")
for (v in c("total_chol", "hdl_chol")) {
  cat(sprintf("  %-11s %.1f%%\n", v, 100 * mi_lambda[[v]]))
}

# Two opposing effects on the mean-imputation standard error, which is why the
# percentages above are close to 100 at this fraction of missing data:
#   1. it ignores the imputation uncertainty just quantified  -> SE too small
#   2. it flattens the variable's spread, and less spread in a
#      predictor means less information about its coefficient  -> SE too large
# The first grows with the fraction missing; the second is the reason the two
# can briefly cancel. Neither makes the SE trustworthy.
cat(sprintf(
  "\nSD of total_chol: %.1f observed -> %.1f after mean imputation\n",
  sd(incomplete$total_chol, na.rm = TRUE), sd(mean_imputed$total_chol)
))

# --- The same comparison with far more missing data ---------------------
# At 12-16% missing the three approaches disagree modestly. Raise the
# missingness to roughly half and the false precision of single imputation
# becomes impossible to miss.
heavy <- punch_holes(framingham, intercept = -0.4)
heavy_mean <- heavy
for (v in c("total_chol", "hdl_chol")) {
  heavy_mean[[v]][is.na(heavy_mean[[v]])] <- mean(heavy_mean[[v]], na.rm = TRUE)
}
imp_h <- mice(heavy, m = 20, method = "pmm", seed = 42, printFlag = FALSE)
pool_h <- pool(with(
  imp_h,
  glm(cvd_10yr ~ age + male + sbp + total_chol + hdl_chol + smoking +
        diabetes + bp_treatment, family = binomial)
))
pooled_h <- summary(pool_h)
lambda_h <- setNames(pool_h$pooled$lambda, pool_h$pooled$term)

cat(sprintf(
  "\n--- With %.0f%% of total_chol and %.0f%% of hdl_chol missing ---\n",
  100 * mean(is.na(heavy$total_chol)), 100 * mean(is.na(heavy$hdl_chol))
))
s_cca_h <- summary(glm(model_formula, data = heavy, family = binomial))$coefficients
s_mean_h <- summary(glm(model_formula, data = heavy_mean, family = binomial))$coefficients
for (v in c("total_chol", "hdl_chol")) {
  cat(sprintf("%-11s truth %+.4f\n", v, truth[[v]]))
  cat(sprintf(
    "  complete case   %+.4f (SE %.5f)\n  mean imputation %+.4f (SE %.5f)\n  MI              %+.4f (SE %.5f)\n",
    s_cca_h[v, 1], s_cca_h[v, 2], s_mean_h[v, 1], s_mean_h[v, 2],
    pooled_h$estimate[pooled_h$term == v],
    pooled_h$std.error[pooled_h$term == v]
  ))
  cat(sprintf(
    "  -> mean imputation's SE is %.0f%% of MI's, and %.0f%% of MI's variance\n     for this coefficient now comes from the imputation (was %.0f%%)\n",
    100 * s_mean_h[v, 2] / pooled_h$std.error[pooled_h$term == v],
    100 * lambda_h[[v]], 100 * mi_lambda[[v]]
  ))
}

cat("
Conclusions
-----------
(a) Complete case analysis is the clear loser. It is the furthest from the
    full-data estimates, and it is biased by construction here: because
    missingness depends on the outcome, the retained rows under-represent
    patients who had an event, and coefficients such as smoking are distorted
    well beyond sampling noise. It also discards a quarter of the cohort, so
    its standard errors are the widest of the three.

    Mean imputation and multiple imputation give similar point estimates at
    this fraction of missing data. Note also that measuring against the true
    values alone is misleading: the diabetes coefficient is far from its true
    0.7 in every column, including the full-data one, because 2000 patients
    and 235 events cannot pin it down. That error is sampling noise, not
    missing-data handling.

(b) Filling every gap with the mean asserts that those values were measured
    rather than guessed, so nothing in the model widens to reflect the
    guessing. Rubin's rules make visible exactly what is being thrown away:
    the printed lambda says what share of the pooled uncertainty comes from
    the imputation itself, and single imputation sets that share to zero.

    That does not always show up as a smaller standard error, and the output
    above is a good reminder to check rather than assume. Two effects pull in
    opposite directions -- ignoring the imputation uncertainty makes the
    standard error too small, while flattening the variable's spread (SD 37.3
    to 35.0) makes it too large -- and at 12-16% missing they nearly cancel,
    leaving mean imputation within 2% of the multiple-imputation standard
    error. Raise the missingness to roughly half and the first effect wins
    outright: mean imputation's standard error for total_chol is about two
    thirds of the honest one.

    Watch the share itself rather than the ratio, because the share is the
    part that behaves predictably: it climbs from under a fifth to two thirds
    or more as the missingness grows. The fair statement is not that single
    imputation always looks more precise, but that its uncertainty is
    unaccounted for, and the size of what it ignores grows with the amount you
    imputed.

    Which method lands nearest the truth in any one dataset is luck; all of
    them sit within a standard error of each other. The missing variance
    component is systematic.

    That is the more serious problem: a biased estimate with an honest
    confidence interval announces its own uncertainty, whereas a spuriously
    precise one invites a confident claim about a coefficient the data cannot
    support. Multiple imputation exists to keep that uncertainty visible.

    One caveat: multiple imputation assumes MAR, which holds here by
    construction. If the highest cholesterol values were missing precisely
    because they were high (MNAR), no method here would recover them, and the
    honest response would be a sensitivity analysis.
")

File: solutions/python/building_prediction_models_ex2.py
"""Exercise 2: Missing data simulation.

Compare complete case analysis, single mean imputation, and multiple
imputation on the simulated Framingham cohort from the chapter.

Python's random numbers are not R's, so the numbers here differ from the R
solution in detail. The pattern -- which approach fails, and how -- is the same.
"""

import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy.special import expit

pd.set_option("display.width", 120)
pd.set_option("display.max_columns", 20)

PREDICTORS = ["age", "male", "sbp", "total_chol", "hdl_chol",
              "smoking", "diabetes", "bp_treatment"]
# The true coefficients: we simulated the data, so we know them.
TRUTH = pd.Series({"const": -7.5, "age": 0.06, "male": 0.4, "sbp": 0.012,
                   "total_chol": 0.005, "hdl_chol": -0.02, "smoking": 0.5,
                   "diabetes": 0.7, "bp_treatment": 0.3})

# --- The cohort, as in the chapter -----------------------------------------
rng = np.random.default_rng(2024)
n = 2000

framingham = pd.DataFrame({
    "age": rng.integers(30, 75, n),
    "male": rng.binomial(1, 0.48, n),
    "sbp": np.round(rng.normal(130, 18, n)),
    "total_chol": np.round(rng.normal(210, 38, n)),
    "hdl_chol": np.round(rng.normal(52, 15, n)),
    "smoking": rng.binomial(1, 0.22, n),
    "diabetes": rng.binomial(1, 0.08, n),
    "bp_treatment": rng.binomial(1, 0.15, n),
})
lp = (-7.5 + 0.06 * framingham["age"] + 0.4 * framingham["male"]
      + 0.012 * framingham["sbp"] + 0.005 * framingham["total_chol"]
      - 0.02 * framingham["hdl_chol"] + 0.5 * framingham["smoking"]
      + 0.7 * framingham["diabetes"] + 0.3 * framingham["bp_treatment"])
framingham["cvd_10yr"] = rng.binomial(1, expit(lp))
print(f"Cohort: {len(framingham)} patients, "
      f"{framingham['cvd_10yr'].sum()} events")


def fit_logit(df):
    """Unpenalised logistic regression, returning coefficients and SEs."""
    X = sm.add_constant(df[PREDICTORS].astype(float))
    res = sm.Logit(df["cvd_10yr"], X).fit(disp=0)
    return res.params, res.bse


def punch_holes(df, intercept, seed=7):
    """Delete values of two predictors under a MAR mechanism.

    Missingness in total_chol depends on the OUTCOME and on age, both of which
    are recorded -- so this is MAR, not MNAR. That is deliberate: if
    missingness depended only on predictors already in the model, complete case
    analysis would still be unbiased for the coefficients and there would be
    nothing to demonstrate.
    """
    out = df.copy()
    r = np.random.default_rng(seed)
    p_chol = expit(intercept + 1.0 * out["cvd_10yr"] + 0.03 * (out["age"] - 52))
    p_hdl = expit(intercept + 0.2 + 0.9 * out["smoking"]
                  + 0.02 * (out["sbp"] - 130))
    out.loc[r.uniform(size=len(out)) < p_chol, "total_chol"] = np.nan
    out.loc[r.uniform(size=len(out)) < p_hdl, "hdl_chol"] = np.nan
    return out


def mean_impute(df):
    out = df.copy()
    for v in ["total_chol", "hdl_chol"]:
        out[v] = out[v].fillna(out[v].mean())
    return out


def impute_once(df, rng, iterations=5):
    """Create ONE filled-in dataset by chained equations.

    This is written out rather than delegated to a package because it is short
    enough to read, and reading it is the point: each missing value is drawn
    from a regression on the other variables, with fresh randomness every time,
    which is what makes the m datasets differ from each other.

    Two details decide whether it works at all:

    * The draw includes both parameter uncertainty (a draw of sigma^2, then of
      beta given sigma^2) and residual noise. Using the fitted value alone
      would make every dataset identical -- single imputation in disguise.
    * The OUTCOME is one of the predictors in the imputation model. Leaving it
      out imputes predictors as though they were unrelated to the outcome,
      which biases their coefficients towards zero. R's `mice` includes every
      column by default, which is why the R solution just hands it the whole
      data frame.

    In practice use `mice` in R or `miceforest` in Python. Note that
    scikit-learn's `IterativeImputer` is not a drop-in substitute: even with
    `sample_posterior=True` its imputations are under-dispersed, which
    understates the between-imputation variance this exercise is about.
    """
    columns = PREDICTORS + ["cvd_10yr"]
    work = df[columns].astype(float).copy()
    incomplete = [v for v in columns if work[v].isna().any()]
    missing = {v: work[v].isna().to_numpy() for v in incomplete}

    for v in incomplete:  # start each variable off at its observed mean
        work.loc[missing[v], v] = work[v].mean()

    for _ in range(iterations):
        for v in incomplete:
            others = [c for c in columns if c != v]
            observed = ~missing[v]
            X = np.column_stack([np.ones(observed.sum()),
                                 work.loc[observed, others].to_numpy()])
            y = work.loc[observed, v].to_numpy()

            XtX_inv = np.linalg.pinv(X.T @ X)
            beta_hat = XtX_inv @ X.T @ y
            residuals = y - X @ beta_hat
            dof = max(len(y) - X.shape[1], 1)

            # Posterior draws, in the order mice's "norm" method uses them
            sigma2 = residuals @ residuals / rng.chisquare(dof)
            chol = np.linalg.cholesky(sigma2 * XtX_inv
                                      + 1e-12 * np.eye(len(beta_hat)))
            beta = beta_hat + chol @ rng.standard_normal(len(beta_hat))

            X_missing = np.column_stack([
                np.ones(missing[v].sum()),
                work.loc[missing[v], others].to_numpy(),
            ])
            work.loc[missing[v], v] = (
                X_missing @ beta
                + rng.normal(0, np.sqrt(sigma2), missing[v].sum())
            )

    work["cvd_10yr"] = df["cvd_10yr"].to_numpy()  # never imputed
    return work


def multiple_impute(df, m=20, seed=42):
    """Impute m times, fit the model in each, pool with Rubin's rules."""
    rng = np.random.default_rng(seed)
    estimates, variances = [], []
    for _ in range(m):
        params, ses = fit_logit(impute_once(df, rng))
        estimates.append(params)
        variances.append(ses**2)

    estimates = pd.DataFrame(estimates)
    variances = pd.DataFrame(variances)

    # Rubin's rules: the pooled variance is the average within-imputation
    # variance plus the between-imputation variance, inflated by 1 + 1/m.
    within = variances.mean()
    between = estimates.var(ddof=1)
    total = within + (1 + 1 / m) * between
    return estimates.mean(), np.sqrt(total), within, between


incomplete = punch_holes(framingham, intercept=-2.2)
complete_rows = incomplete[PREDICTORS].notna().all(axis=1)

print(f"Missing total_chol: {incomplete['total_chol'].isna().sum()}")
print(f"Missing hdl_chol:  {incomplete['hdl_chol'].isna().sum()}")
print(f"Complete rows: {complete_rows.sum()} "
      f"({100 * (1 - complete_rows.mean()):.0f}% of the cohort discarded by a "
      "complete case analysis)")

# --- The three approaches, plus the full data as a benchmark -------------
b_full, se_full = fit_logit(framingham)
b_cca, se_cca = fit_logit(incomplete[complete_rows])
b_mean, se_mean = fit_logit(mean_impute(incomplete))
b_mi, se_mi, within, between = multiple_impute(incomplete)

comparison = pd.DataFrame({
    "truth": TRUTH,
    "full_data": b_full,
    "complete_case": b_cca,
    "mean_imputation": b_mean,
    "multiple_imputation": b_mi,
})[["truth", "full_data", "complete_case", "mean_imputation",
    "multiple_imputation"]]

print("\n--- Coefficient estimates ---")
print(comparison.round(4))

# --- (a) Which approach lands closest to the truth? ---------------------
# Two error measures, and the difference between them is the point. Distance
# from the truth mixes up two things: damage done by the missing data, and the
# sampling noise already present in this cohort of 2000. Distance from the
# full-data estimates isolates the first.
slopes = comparison.drop(index="const")
errors = pd.DataFrame({
    "vs_truth": (slopes.drop(columns="truth")
                 .sub(slopes["truth"], axis=0).abs().mean()),
    "vs_full_data": (slopes.drop(columns=["truth", "full_data"])
                     .sub(slopes["full_data"], axis=0).abs().mean()),
})
print("\n--- Mean absolute error across the 8 slopes ---")
print(errors.round(5))

print("\nRanked by distance from the full-data estimates:")
print(errors["vs_full_data"].dropna().sort_values().round(5))

# --- (b) What happens to the standard errors? --------------------------
se_table = pd.DataFrame({
    "complete_case": se_cca,
    "mean_imputation": se_mean,
    "multiple_imputation": se_mi,
})
print("\n--- Standard errors ---")
print(se_table.round(5))

print("\nMean imputation SE as a percentage of the multiple-imputation SE,")
print("for the two variables that were actually imputed:")
for v in ["total_chol", "hdl_chol"]:
    print(f"  {v:11s} {100 * se_mean[v] / se_mi[v]:.1f}%")

# What single imputation actually discards, as a number. Rubin's rules split the
# pooled variance into a within-imputation part (ordinary sampling uncertainty)
# and a between-imputation part (uncertainty about the guesses themselves).
# Single imputation sets the second to zero by construction. This share is what
# mice reports as lambda.
print("\nShare of the pooled MI variance that comes from the imputation itself:")
for v in ["total_chol", "hdl_chol"]:
    share = (1 + 1 / 20) * between[v] / (within[v] + (1 + 1 / 20) * between[v])
    print(f"  {v:11s} {100 * share:.1f}%")

# Two opposing effects act on the mean-imputation standard error, which is why
# the percentages above sit close to 100 at this fraction of missing data:
#   1. it ignores the imputation uncertainty just quantified  -> SE too small
#   2. it flattens the variable's spread, and less spread in a
#      predictor means less information about its coefficient  -> SE too large
# The first grows with the fraction missing; the second is why they can briefly
# cancel, and one of the two variables below may even come out above 100%.
print(f"\nSD of total_chol: {incomplete['total_chol'].std():.1f} observed"
      f" -> {mean_impute(incomplete)['total_chol'].std():.1f}"
      " after mean imputation")

# --- The same comparison with far more missing data -------------------
heavy = punch_holes(framingham, intercept=-0.4)
heavy_rows = heavy[PREDICTORS].notna().all(axis=1)
b_cca_h, se_cca_h = fit_logit(heavy[heavy_rows])
b_mean_h, se_mean_h = fit_logit(mean_impute(heavy))
b_mi_h, se_mi_h, within_h, between_h = multiple_impute(heavy)

print(f"\n--- With {100 * heavy['total_chol'].isna().mean():.0f}% of total_chol"
      f" and {100 * heavy['hdl_chol'].isna().mean():.0f}% of hdl_chol missing ---")
for v in ["total_chol", "hdl_chol"]:
    print(f"{v:11s} truth {TRUTH[v]:+.4f}")
    print(f"  complete case   {b_cca_h[v]:+.4f} (SE {se_cca_h[v]:.5f})")
    print(f"  mean imputation {b_mean_h[v]:+.4f} (SE {se_mean_h[v]:.5f})")
    print(f"  MI              {b_mi_h[v]:+.4f} (SE {se_mi_h[v]:.5f})")
    share_now = ((1 + 1 / 20) * between_h[v]
                 / (within_h[v] + (1 + 1 / 20) * between_h[v]))
    share_was = ((1 + 1 / 20) * between[v]
                 / (within[v] + (1 + 1 / 20) * between[v]))
    print(f"  -> mean imputation's SE is "
          f"{100 * se_mean_h[v] / se_mi_h[v]:.0f}% of MI's, and "
          f"{100 * share_now:.0f}% of MI's variance")
    print(f"     for this coefficient now comes from the imputation "
          f"(was {100 * share_was:.0f}%)")

print("""
Conclusions
-----------
(a) Complete case analysis is the clear loser. It sits furthest from the
    full-data estimates, and it is biased by construction here: because
    missingness depends on the outcome, the retained rows under-represent
    patients who had an event. It also discards a quarter of the cohort, so
    its standard errors are the widest of the three.

    Mean imputation and multiple imputation give similar point estimates at
    this fraction of missing data. Measuring against the true values alone is
    misleading, though: the diabetes coefficient is far from its true 0.7 in
    every column, including the full-data one, because 2000 patients and
    roughly 200 events cannot pin it down. That error is sampling noise, not
    missing-data handling -- which is why the comparison against the full-data
    estimates is the informative one.

(b) Filling every gap with the mean asserts those values were measured rather
    than guessed, so nothing in the model widens to reflect the guessing.
    Rubin's rules make visible exactly what is discarded: the printed share
    says how much of the pooled uncertainty comes from the imputation itself,
    and single imputation sets that share to zero.

    That does not always surface as a smaller standard error, and the output
    above is a good reminder to check rather than assume. Two effects pull in
    opposite directions -- ignoring the imputation uncertainty makes the
    standard error too small, while flattening the variable's spread makes it
    too large -- so at this fraction of missing data they nearly cancel, and
    with half the values missing one variable still comes out slightly above
    the multiple-imputation standard error while the other falls to about two
    thirds of it.

    Watch the share rather than the ratio, because the share is the part that
    behaves predictably: it climbs from a fifth or a quarter to a half or more
    as the missingness grows. The fair statement is not that single imputation always
    looks more precise, but that its uncertainty is unaccounted for, and the
    size of what it ignores grows with the amount you imputed.

    Which method lands nearest the truth in any one dataset is luck; all of
    them sit within a standard error of each other. The missing variance
    component is systematic.

    One caveat: multiple imputation assumes MAR, which holds here by
    construction. If the highest cholesterol values were missing precisely
    because they were high (MNAR), no method here would recover them, and the
    honest response would be a sensitivity analysis.
""")
